File: module_1/python/reverse_list.py
"""
Title: Reverse Linked List
Leetcode Link: https://leetcode.com/problems/reverse-linked-list/

Problem: Given a singly linked list, reverse the list.

Input:
    ListNode head   => The head of the linked list
Output:
    ListNode        => The head of the reversed list

Execution: python reverse_list.py
"""
import unittest

# Basic node class
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

"""
Solution #1: Brute Force

In this solution, we add all the nodes to an builtin list, allowing us to easily
maintain order while swapping pointers.

This is not recommended and for demonstrative purposes only.

Time Complexity: O(n)
Space Complexity: O(n)
"""
def reverse_list_bf(head: ListNode) -> ListNode:
    if not head:
        return head
    # Create an list that contains references to all the list nodes
    nodes = []
    curr = head
    while curr:
        nodes.append(curr)
        curr = curr.next

    # Reverse pointers
    nodes[0].next = None
    for i in range(1, len(nodes)):
        nodes[i].next = nodes[i-1]

    return nodes[-1]

"""
Solution #2: Iterative

Iterate over all the nodes and reverse pointers as we go

Time Complexity: O(n)
Space Complexity: O(1)
"""
def reverse_list_iter(head: ListNode) -> ListNode:
    prev = None
    curr = head

    # Iterate over the list and swap pointers as we go
    while curr:
        next_temp = curr.next
        curr.next = prev
        prev = curr
        curr = next_temp
    return prev

"""
Solution #3: Recursive

Recursively reverse the remainder of the list and point current node to the
previous

Time Complexity: O(n)
Space Complexity: O(n)
"""
def reverse_list_rec(head: ListNode) -> ListNode:
    # If we're at the end of the list, just return head
    if head is None or head.next is None:
        return head

    # Reverse the list from head.next to the end
    rem = reverse_list_rec(head.next)

    # Point the last node of that reversed list to the current node
    head.next.next = head
    head.next = None
    return rem

class TestReverseList(unittest.TestCase):
    """Unit test for reverse list."""

    def test_1(self):
        """Test for 1->2->3->4->5."""
        input_1 = ListNode(1)
        input_1.next = ListNode(2)
        input_1.next.next = ListNode(3)
        input_1.next.next.next = ListNode(4)
        input_1.next.next.next.next = ListNode(5)

        output_1 = reverse_list_rec(input_1)
        l = []
        while output_1:
            l.append(output_1.val)
            output_1 = output_1.next

        self.assertEqual(l, [5, 4, 3, 2, 1])

    # ADD YOUR TESTS HERE

if __name__ == '__main__':
    unittest.main()

File: module_1/javascript/ReverseList.js
/*
 *   Title: Reverse Linked List
 *   Leetcode Link: https://leetcode.com/problems/reverse-linked-list/
 *
 *   Problem: Given a singly linked list, reverse the list.
 *
 *   Input:
 *      ListNode head   => The head of the linked list
 *   Output:
 *      ListNode        => The head of the reversed list
 *
 *   Execution: node ReverseList.js
 */

 /*
  * Basic node class
  */
class ListNode {
    constructor(val, next) {
        this.val = val;
        this.next = next;
    }
}

/**
 * Solution #1: Brute Force
 *
 * In this solution, we add all the nodes to an array, allowing us to easily
 * maintain order while swapping pointers.
 *
 * This is not recommended and for demonstrative purposes only.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * @param {ListNode} head
 * @return {ListNode}
 */
 var reverseListBF = function(head) {
     if (!head) return head;

     // Create an array that contains references to all the list nodes
     var curr = head;
     var nodes = [];
     while (curr) {
         nodes.push(curr);
         curr = curr.next;
     }

     // Reverse pointers
     nodes[0].next = null;
     for (var i = 1; i < nodes.length; i++) {
         nodes[i].next = nodes[i-1];
     }

     return nodes[nodes.length - 1];
 }

 /**
  * Solution #2: Iterative
  *
  * Iterate over all the nodes and reverse pointers as we go
  *
  * Time Complexity: O(n)
  * Space Complexity: O(1)
  *
  * @param {ListNode} head
  * @return {ListNode}
  */
var reverseListIter = function(head) {
    var prev = null;
    var curr = head;

    // Iterate over the list and swap pointers as we go
    while (curr != null) {
        var nextTemp = curr.next;
        curr.next = prev;
        prev = curr;
        curr = nextTemp;
    }
    return prev;
}

/**
 * Solution #3: Recursive
 *
 * Recursively reverse the remainder of the list and point current node to
 * the previous
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseListRec = function(head) {
    // If we're at the end of the list, just return head
    if (!head || !head.next) {
        return head;
    }

    // Reverse the list from head.next to the end
    var rem = reverseListRec(head.next);

    // Point the last node of that reversed list to the current node
    head.next.next = head;
    head.next = null;
    return rem;
}

var tester = function() {
    var l = new ListNode(1);
    l.next = new ListNode(2);
    l.next.next = new ListNode(3);
    l.next.next.next = new ListNode(4);

    var r = reverseListRec(l);
    while (r) {
        console.log(r.val);
        r = r.next;
    }

    // ADD YOUR TESTS HERE
}

tester();
